src/state_machine.zig

```diff
@@ -337,151 +331,161 @@ pub const StateMachine = struct {
     }
 
     fn create_transfer(self: *StateMachine, t: Transfer) CreateTransferResult {
-        assert(t.timestamp > self.commit_timestamp);
-
-        if (t.flags.padding != 0) return .reserved_flag_padding;
-        if (t.flags.two_phase_commit) {
-            // Otherwise reserved amounts may never be released:
-            if (t.timeout == 0) return .two_phase_commit_must_timeout;
-        } else if (t.timeout != 0) {
-            return .timeout_reserved_for_two_phase_commit;
-        }
-        if (!t.flags.condition and !zeroed_32_bytes(t.reserved)) return .reserved_field;
+        assert(t.timestamp > self.post_timestamp);
 
-        if (t.amount == 0) return .amount_is_zero;
+        // Either a 2-phase transfer post/void
+        if (t.flags.post_pending_transfer and t.flags.void_pending_transfer) {
+            return .cannot_void_and_post_two_phase_commit;
+        } else if (t.flags.post_pending_transfer or t.flags.void_pending_transfer) {
+            if (!t.flags.hashlock and !zeroed_32_bytes(t.reserved)) return .reserved_field;
+            if (t.flags.padding != 0) return .reserved_flag_padding;
 
-        if (t.debit_account_id == t.credit_account_id) return .accounts_are_the_same;
+            var lookup = self.get_transfer(t.id) orelse return .transfer_not_found;
+            assert(t.timestamp > lookup.timestamp);
 
-        // The etymology of the DR and CR abbreviations for debit and credit is interesting, either:
-        // 1. derived from the Latin past participles of debitum and creditum, debere and credere,
-        // 2. standing for debit record and credit record, or
-        // 3. relating to debtor and creditor.
-        // We use them to distinguish between `cr` (credit account), and `c` (commit).
-        var dr = self.get_account(t.debit_account_id) orelse return .debit_account_not_found;
-        var cr = self.get_account(t.credit_account_id) orelse return .credit_account_not_found;
-        assert(t.timestamp > dr.timestamp);
-        assert(t.timestamp > cr.timestamp);
+            if (!lookup.flags.pending) return .transfer_not_two_phase_commit;
 
-        if (dr.unit != cr.unit) return .accounts_have_different_units;
+            if (self.get_commit(t.id)) |exists| {
+                if (!exists.flags.void_pending_transfer and t.flags.void_pending_transfer) return .already_committed_but_accepted;
+                if (exists.flags.void_pending_transfer and !t.flags.void_pending_transfer) return .already_committed_but_rejected;
+                return .already_committed;
+            }
 
-        // TODO We need a lookup before inserting in case transfer exists and would overflow limits.
```

Comment:
@sentientwaffle The logic has been updated to perform the `Transfer` lookup/validation prior to exceeding limit checks.
Once the `exists` check is performed, the `{debits/credits}_exceed_{credits/debits}` is performed. 
The `{debits/credits}_exceed_{credits/debits}` verification performs a validation on the sum of the `posted` and `pending` against the account `credits_posted`/`debits_posted`. 
> So no need to check `{credits,debits}_pending`? **Unless I am missing something?**

```zig
// 1. Ensure the transfer is not duplicated prior to limit verification's.
const insert = self.transfers.getOrPutAssumeCapacity(t.id);
            if (insert.found_existing) {
                const exists = insert.value_ptr.*;
                if (exists.debit_account_id != t.debit_account_id) {
                    return .exists_with_different_debit_account_id;
                } else if (exists.credit_account_id != t.credit_account_id) {
                    return .exists_with_different_credit_account_id;
                }
                if (exists.amount != t.amount) return .exists_with_different_amount;
                if (@bitCast(u32, exists.flags) != @bitCast(u32, t.flags)) {
                    return .exists_with_different_flags;
                }
                if (exists.user_data != t.user_data) return .exists_with_different_user_data;
                if (!equal_32_bytes(exists.reserved, t.reserved)) {
                    return .exists_with_different_reserved_field;
                }
                if (exists.timeout != t.timeout) return .exists_with_different_timeout;
                return .exists;
            }

// 2. Verify limits, if limits are exceeded, the transfer is removed. 
            if (dr.debits_exceed_credits(t.amount)) {
                assert(self.transfers.remove(t.id));
                return .exceeds_credits;
            } else if (cr.credits_exceed_debits(t.amount)) {
                assert(self.transfers.remove(t.id));
                return .exceeds_debits;
            }

// 3. Lastly we increment the amounts
insert.value_ptr.* = t;
            if (t.flags.pending) {
                dr.debits_pending += t.amount;
                cr.credits_pending += t.amount;
            } else {
                dr.debits_posted += t.amount;
                cr.credits_posted += t.amount;
            }
            self.commit_timestamp = t.timestamp;
            return .ok;
```

Reply:
Maybe I misunderstood what this `TODO` was referring to. But what I mean is that we need to verify that `self.debits_pending + self.debits_posted + amount` (and likewise for `credits`) doesn't overflow a `u64`, since that would panic.

Reply:
@sentientwaffle You are right! :bowing_man:  I have updated the `tigerbeetle.zig` to handle the overflow. I have also added a test case for it `test "exceed credit/debit limits panic"`.